File: src/moltblock/handoff.py
"""Multi-entity handoff: Entity A produces signed artifact -> Entity B consumes as input."""

from .persistence import Store, get_inbox, put_inbox
from .signing import artifact_hash, sign_artifact, verify_artifact


def send_artifact(
    sender_entity_id: str,
    recipient_store: Store,
    artifact_content: str,
    artifact_ref: str | None = None,
) -> str:
    """
    Sign artifact and deliver to recipient's inbox. Returns artifact_ref.
    recipient_store is the recipient entity's Store (entity_id = recipient).
    """
    import time
    ref = artifact_ref or f"artifact_{sender_entity_id}_{int(time.time() * 1000)}"
    payload_hash = artifact_hash(artifact_content)
    signature = sign_artifact(sender_entity_id, artifact_content)
    put_inbox(
        recipient_store,
        from_entity_id=sender_entity_id,
        artifact_ref=ref,
        payload_hash=payload_hash,
        signature=signature,
        payload_text=artifact_content[:100_000],
    )
    return ref


def receive_artifacts(
    store: Store,
    limit: int = 20,
    verify: bool = True,
) -> list[dict]:
    """
    Get inbox artifacts for this entity. If verify=True, only return entries
    where the signature is valid for the sender. Each entry includes
    from_entity_id, artifact_ref, payload_text, verified.
    """
    entries = get_inbox(store, limit=limit)
    result = []
    for e in entries:
        ok = True
        if verify and e.get("payload_text") and e.get("signature"):
            ok = verify_artifact(e["from_entity_id"], e["payload_text"], e["signature"])
        result.append({
            "from_entity_id": e["from_entity_id"],
            "artifact_ref": e["artifact_ref"],
            "payload_text": e.get("payload_text", ""),
            "verified": ok,
        })
    return result

File: src/handoff.ts
/**
 * Multi-entity handoff: Entity A produces signed artifact -> Entity B consumes as input.
 */

import { Store, getInbox, putInbox } from "./persistence.js";
import { artifactHash, signArtifact, verifyArtifact } from "./signing.js";
import type { ReceivedArtifact } from "./types.js";

/**
 * Sign artifact and deliver to recipient's inbox. Returns artifact_ref.
 * recipientStore is the recipient entity's Store (entity_id = recipient).
 */
export function sendArtifact(
  senderEntityId: string,
  recipientStore: Store,
  artifactContent: string,
  artifactRef?: string
): string {
  const ref = artifactRef ?? `artifact_${senderEntityId}_${Date.now()}`;
  const payloadHash = artifactHash(artifactContent);
  const signature = signArtifact(senderEntityId, artifactContent);

  putInbox(
    recipientStore,
    senderEntityId,
    ref,
    payloadHash,
    signature,
    artifactContent.slice(0, 100_000)
  );

  return ref;
}

/**
 * Get inbox artifacts for this entity. If verify=true, only return entries
 * where the signature is valid for the sender. Each entry includes
 * from_entity_id, artifact_ref, payload_text, verified.
 */
export function receiveArtifacts(
  store: Store,
  options: { limit?: number; verify?: boolean } = {}
): ReceivedArtifact[] {
  const { limit = 20, verify = true } = options;
  const entries = getInbox(store, limit);

  return entries.map((e) => {
    let ok = true;
    if (verify && e.payload_text && e.signature) {
      ok = verifyArtifact(e.from_entity_id, e.payload_text, e.signature);
    }
    return {
      from_entity_id: e.from_entity_id,
      artifact_ref: e.artifact_ref,
      payload_text: e.payload_text ?? "",
      verified: ok,
    };
  });
}
